File: apps/web/lib/types/donor.ts
export type DonorType = "egg" | "sperm"
export type DonorPipelineEntityType = "egg_donor" | "sperm_donor"
export type DonorOwnerType = "user" | "queue"

export interface Donor {
    id: string
    donor_number: string
    donor_type: DonorType
    full_name: string
    email: string
    phone: string | null
    state: string | null
    education: string | null
    source: string | null
    owner_type: DonorOwnerType | null
    owner_id: string | null
    stage_id: string
    stage_key: string
    stage_slug: string
    status: string
    status_label: string
    profile_photo_attachment_id: string | null
    is_archived: boolean
    archived_at: string | null
    created_at: string
    updated_at: string
}

export type DonorListItem = Donor

export interface DonorListResponse {
    items: DonorListItem[]
    total: number
    page: number
    per_page: number
    pages: number
}

export interface DonorCreate {
    donor_type: DonorType
    full_name: string
    email: string
    phone?: string | null
    state?: string | null
    education?: string | null
    source?: string | null
    owner_type?: DonorOwnerType | null
    owner_id?: string | null
}

export type DonorUpdate = Partial<Omit<DonorCreate, "donor_type">>

export interface DonorStatusUpdate {
    stage_id: string
    reason?: string
    effective_at?: string
}

export interface DonorStatusHistoryItem {
    id: string
    donor_id: string
    changed_by_user_id: string | null
    changed_by_name: string | null
    old_stage_id: string | null
    new_stage_id: string | null
    old_status: string | null
    new_status: string
    old_label_snapshot: string | null
    new_label_snapshot: string
    reason: string | null
    effective_at: string
    recorded_at: string
    requested_at: string | null
    approved_by_user_id: string | null
    approved_by_name: string | null
    approved_at: string | null
    is_undo: boolean
    request_id: string | null
}

export interface DonorStatusChangeResponse {
    status: "applied" | "pending_approval"
    donor: Donor | null
    history: DonorStatusHistoryItem | null
    request_id: string | null
    message: string | null
}

export interface DonorNoteListItem {
    id: string
    author_id: string
    content: string
    created_at: string
}

export interface DonorNote extends DonorNoteListItem {
    organization_id: string
    entity_type: "donor"
    entity_id: string
}

export interface DonorNoteCreate {
    content: string
}

export function getDonorTypeLabel(donorType: DonorType): string {
    return donorType === "egg" ? "Egg Donor" : "Sperm Donor"
}

export function getDonorTypePluralLabel(donorType: DonorType): string {
    return donorType === "egg" ? "Egg Donors" : "Sperm Donors"
}

export function getDonorPipelineEntityType(donorType: DonorType): DonorPipelineEntityType {
    return donorType === "egg" ? "egg_donor" : "sperm_donor"
}

File: apps/api/app/schemas/donor.py
"""Pydantic contracts for donors."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from app.utils.normalization import normalize_phone, normalize_state

DonorTypeValue = Literal["egg", "sperm"]


class DonorCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    donor_type: DonorTypeValue
    full_name: str = Field(min_length=1, max_length=255)
    email: EmailStr
    phone: str | None = Field(None, max_length=50)
    state: str | None = Field(None, max_length=100)
    education: str | None = Field(None, max_length=255)
    source: str | None = Field(None, max_length=100)
    owner_type: Literal["user", "queue"] | None = None
    owner_id: UUID | None = None

    @field_validator("phone", mode="before")
    @classmethod
    def normalize_phone_field(cls, value: str | None) -> str | None:
        return normalize_phone(value) if value else None

    @field_validator("state", mode="before")
    @classmethod
    def normalize_state_field(cls, value: str | None) -> str | None:
        return normalize_state(value)


class DonorUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    full_name: str | None = Field(None, min_length=1, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(None, max_length=50)
    state: str | None = Field(None, max_length=100)
    education: str | None = Field(None, max_length=255)
    source: str | None = Field(None, max_length=100)
    owner_type: Literal["user", "queue"] | None = None
    owner_id: UUID | None = None

    @field_validator("phone", mode="before")
    @classmethod
    def normalize_phone_field(cls, value: str | None) -> str | None:
        return normalize_phone(value) if value else None

    @field_validator("state", mode="before")
    @classmethod
    def normalize_state_field(cls, value: str | None) -> str | None:
        return normalize_state(value)


class DonorStatusUpdate(BaseModel):
    stage_id: UUID
    reason: str | None = Field(None, max_length=2000)
    effective_at: datetime | None = Field(
        None, description="When the change actually occurred (optional, defaults to now)"
    )


class DonorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    donor_number: str
    donor_type: DonorTypeValue
    full_name: str
    email: str
    phone: str | None
    state: str | None
    education: str | None
    source: str | None
    owner_type: str | None
    owner_id: UUID | None
    stage_id: UUID
    status: str
    stage_key: str
    stage_slug: str
    status_label: str
    profile_photo_attachment_id: UUID | None
    is_archived: bool
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class DonorListResponse(BaseModel):
    items: list[DonorRead]
    total: int
    page: int
    per_page: int
    pages: int


class DonorStatusHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    donor_id: UUID
    changed_by_user_id: UUID | None
    changed_by_name: str | None = None
    old_stage_id: UUID | None
    new_stage_id: UUID | None
    old_status: str | None
    new_status: str
    old_label_snapshot: str | None
    new_label_snapshot: str
    reason: str | None
    effective_at: datetime
    recorded_at: datetime
    requested_at: datetime | None = None
    approved_by_user_id: UUID | None = None
    approved_by_name: str | None = None
    approved_at: datetime | None = None
    is_undo: bool = False
    request_id: UUID | None = None


class DonorStatusChangeResponse(BaseModel):
    status: Literal["applied", "pending_approval"]
    donor: DonorRead | None = None
    history: DonorStatusHistoryRead | None = None
    request_id: UUID | None = None
    message: str | None = None
